Spigot-API-Patches/0167-Add-BaseComponent-array-to-a-variety-of-events.patch

```diff
@@ -0,0 +1,2079 @@
++    }
++
++    private static boolean isValidFormatLegacyText(final FormatSpecifier format, final String text, final int off, final int arguments) {
++        for (int nextFormat = text.indexOf('%', off); nextFormat != -1 && nextFormat < text.length();) {
++            nextFormat = FormatSpecifier.readFormatSpecifier(format, text, nextFormat, arguments);
++            if (nextFormat == -1) {
++                return false;
++            }
++        }
++        return true;
++    }
++
++    /**
++     * Converts the BaseComponent format to legacy format. The conversion is not 1:1, below lists the changes:
++     * <ul>
++     *     <li>
++     *         All of the formatting flags are removed for each specifier.
++     *     </li>
++     *     <li>
++     *         The ordering of the flags is not retained.
++     *     </li>
++     *     <li>
++     *         Any BaseComponent specific feature is removed.
++     *     </li>
++     *     <li>
++     *         Width and precision arguments are not guaranteed to be interpreted as originally intended.
++     *     </li>
++     * </ul>
++     * @param format The specified format to convert.
++     * @throws IllegalArgumentException If the specified format is invalid as determined by {@link #isValidFormat(BaseComponent[], int)}, but with arguments = {@link Integer#MAX_VALUE}
++     * @return The converted format.
++     */
++    public static String convertToLegacyFormat(final BaseComponent[] format) throws IllegalArgumentException {
++        /* For reference: %[argument_index$][flags][width][.precision]conversion */
++        /* Note: Flag order does not matter */
++        if (!BaseComponentFormatter.isValidFormat(format, Integer.MAX_VALUE)) {
++            /* Catch invalid formatting that would be across BaseComponents */
++            /* i.e {TextComponent("%"), TextComponent("s")} */
++            throw new IllegalArgumentException();
++        }
++        final StringBuilder legacyFormat = new StringBuilder(64);
++        final FormatSpecifier specifier = new FormatSpecifier();
++
++        final String legacyText = TextComponent.toLegacyText(format);
++
++        int off, lastOff = 0;
++        for (off = legacyText.indexOf('%'); off != -1;) {
++            legacyFormat.append(legacyFormat, lastOff, off);
++
++            final int next = FormatSpecifier.readFormatSpecifier(specifier, legacyText, off, Integer.MAX_VALUE);
++            if (next == -1) {
++                throw new IllegalArgumentException();
++            }
++            lastOff = next; /* skip the specifier */
++
++            off = legacyText.indexOf('%', lastOff);
++
++            legacyFormat.append('%');
++
++            if (specifier.conversion == 'n' || specifier.conversion == '%') {
++                legacyFormat.append(specifier.conversion);
++                continue;
++            }
++
++            legacyFormat.append(specifier.argumentIndex).append('$');
```

Comment:
note why explicit argument index is required.

Reply:
nevermind this function is broken as it does not account for implicit arguments correctly

Reply:
now that order is defined to be from right to left in the string this does not need to be fixed (order elsewhere is undefined)